### PR TITLE
Add splunk_generic_acl resource

### PR DESCRIPTION
**Pull request by micahkemp-splunk**

### CODE REVIEW

File: client/acl.go
Comment (micahkemp-splunk) [inline client/acl.go:14]: This seemed to be incorrect in general, and caused issues trying to use it without changing, because it has owner and app, which is only valid for `servicesNS`.

This method didn't seem to be used anywhere else, based on a simple `grep`.

File: client/acl.go
Comment (micahkemp-splunk) [inline client/acl.go:20]: I changed this while troubleshooting some failures, but I think more detailed errors will be helpful.

**Comment by micahkemp-splunk** [comment]
Before this is ready to be merged, I need to add testing and documentation.  I'm holding off on adding those until getting feedback on how this was implemented.

File: splunk/provider.go
Comment (micahkemp-splunk) [inline splunk/provider.go:79]: I chose this name to be similar to how the Vault provider names some resources. This resource could be used for pretty much anything you want to apply an ACL on, but as such it requires a bit of thinking to do so properly.

My thought is I would also add a `splunk_app_acl`, `splunk_view_acl`, etc. for higher frequency use cases which would be more user friendly.

**Comment by anushjay** [comment]
@micahkemp-splunk 
I'm starting to review your changes, though this helps to add ACL to any resources I'm afraid that this might cause conflict with resources that were created with some kind of default ACL, to begin with. For instance, what happens to the ACL of the resource created with `splunk_data_ui_views` after changing it using the generic_acl resource. Does the `splunk_data_ui_views` resource get updated? 🤔

**Comment by micahkemp-splunk** [comment]
> @micahkemp-splunk
> I'm starting to review your changes, though this helps to add ACL to any resources I'm afraid that this might cause conflict with resources that were created with some kind of default ACL, to begin with. For instance, what happens to the ACL of the resource created with `splunk_data_ui_views` after changing it using the generic_acl resource. Does the `splunk_data_ui_views` resource get updated? 🤔

This resource is intended to be used only for resources that aren't already managed in terraform.  Specifically it's for remote knowledge objects that can't or won't be managed in terraform in terms of content, but would still be preferable to manage their permissions in terraform.

Think about things like uploaded apps or dashboards that come in an app.  The app may not be installable via terraform (such as custom apps not hosted on splunkbase), or a dashboard may come with an app, but we want to manage its permissions in terraform (but not its content).

**Comment by anushjay** [comment]
> > @micahkemp-splunk
> > I'm starting to review your changes, though this helps to add ACL to any resources I'm afraid that this might cause conflict with resources that were created with some kind of default ACL, to begin with. For instance, what happens to the ACL of the resource created with `splunk_data_ui_views` after changing it using the generic_acl resource. Does the `splunk_data_ui_views` resource get updated? 🤔
> 
> This resource is intended to be used only for resources that aren't already managed in terraform. Specifically it's for remote knowledge objects that can't or won't be managed in terraform in terms of content, but would still be preferable to manage their permissions in terraform.
> 
> Think about things like uploaded apps or dashboards that come in an app. The app may not be installable via terraform (such as custom apps not hosted on splunkbase), or a dashboard may come with an app, but we want to manage its permissions in terraform (but not its content).

For existing resources that are being managed by terraform, executing `terraform refresh` command after changing the ACL of the resource using the `generic_acl` should update the original resource as well. Can we add this as a *note* in the readme?

File: client/acl.go
Comment (anushjay) [reply client/acl.go:14]: Ok. Yeah, I guess we never ran into errors since it was not being used. `servicesNS` makes sense.

**Comment by micahkemp-splunk** [comment]
> > > @micahkemp-splunk
> > > I'm starting to review your changes, though this helps to add ACL to any resources I'm afraid that this might cause conflict with resources that were created with some kind of default ACL, to begin with. For instance, what happens to the ACL of the resource created with `splunk_data_ui_views` after changing it using the generic_acl resource. Does the `splunk_data_ui_views` resource get updated? 🤔
> > 
> > 
> > This resource is intended to be used only for resources that aren't already managed in terraform. Specifically it's for remote knowledge objects that can't or won't be managed in terraform in terms of content, but would still be preferable to manage their permissions in terraform.
> > Think about things like uploaded apps or dashboards that come in an app. The app may not be installable via terraform (such as custom apps not hosted on splunkbase), or a dashboard may come with an app, but we want to manage its permissions in terraform (but not its content).
> 
> For existing resources that are being managed by terraform, executing `terraform refresh` command after changing the ACL of the resource using the `generic_acl` should update the original resource as well. Can we add this as a _note_ in the readme?

Would it make more sense to add a note *not* to use `generic_acl` for objects already managed in terraform, but to instead use that resource's `acl` configuration instead?

When you say "should update the original resource" that's only the state.  The configuration and the state would have drift, and an additional `apply` would be problematic with two resources attempting to manage the same object with two different configurations.

**Review by anushjay — COMMENTED** [review]
> > > > @micahkemp-splunk
> > > > I'm starting to review your changes, though this helps to add ACL to any resources I'm afraid that this might cause conflict with resources that were created with some kind of default ACL, to begin with. For instance, what happens to the ACL of the resource created with `splunk_data_ui_views` after changing it using the generic_acl resource. Does the `splunk_data_ui_views` resource get updated? 🤔
> > > 
> > > 
> > > This resource is intended to be used only for resources that aren't already managed in terraform. Specifically it's for remote knowledge objects that can't or won't be managed in terraform in terms of content, but would still be preferable to manage their permissions in terraform.
> > > Think about things like uploaded apps or dashboards that come in an app. The app may not be installable via terraform (such as custom apps not hosted on splunkbase), or a dashboard may come with an app, but we want to manage its permissions in terraform (but not its content).
> > 
> > 
> > For existing resources that are being managed by terraform, executing `terraform refresh` command after changing the ACL of the resource using the `generic_acl` should update the original resource as well. Can we add this as a _note_ in the readme?
> 
> Would it make more sense to add a note _not_ to use `generic_acl` for objects already managed in terraform, but to instead use that resource's `acl` configuration instead?
> 
> When you say "should update the original resource" that's only the state. The configuration and the state would have drift, and an additional `apply` would be problematic with two resources attempting to manage the same object with two different configurations.

Yes, that is correct. `refresh` should only update the state

> > > > @micahkemp-splunk
> > > > I'm starting to review your changes, though this helps to add ACL to any resources I'm afraid that this might cause conflict with resources that were created with some kind of default ACL, to begin with. For instance, what happens to the ACL of the resource created with `splunk_data_ui_views` after changing it using the generic_acl resource. Does the `splunk_data_ui_views` resource get updated? 🤔
> > > 
> > > 
> > > This resource is intended to be used only for resources that aren't already managed in terraform. Specifically it's for remote knowledge objects that can't or won't be managed in terraform in terms of content, but would still be preferable to manage their permissions in terraform.
> > > Think about things like uploaded apps or dashboards that come in an app. The app may not be installable via terraform (such as custom apps not hosted on splunkbase), or a dashboard may come with an app, but we want to manage its permissions in terraform (but not its content).
> > 
> > 
> > For existing resources that are being managed by terraform, executing `terraform refresh` command after changing the ACL of the resource using the `generic_acl` should update the original resource as well. Can we add this as a _note_ in the readme?
> 

> When you say "should update the original resource" that's only the state. The configuration and the state would have drift, and an additional `apply` would be problematic with two resources attempting to manage the same object with two different configurations.

Yes, `refresh` should only update the state. But an `apply` is not required at that point. I feel it is better to keep the state file in sync. (Not required I believe)

> Would it make more sense to add a note _not_ to use `generic_acl` for objects already managed in terraform, but to instead use that resource's `acl` configuration instead?
> 

Yeah, that sounds like a better idea. Adding a suggestion like this? 
```
Use `generic_acl` to manage ACL of Splunk objects that are not managed as resources by Terraform.
```

File: splunk/provider.go
Comment (anushjay) [reply splunk/provider.go:79]: Yeah, sounds good.

**Review by anushjay — APPROVED** [review]

**Pull request merged**
